R/build_lightgbm.R

```diff
@@ -312,6 +318,124 @@ fml_lightgbm <- function(data, formula, nrounds = 10, weights = NULL, watchlist_
     callbacks <- c(callbacks, list(lightgbm::cb.record.evaluation(evals_result)))
   }
 
+  # Cache xlevels so we can build consistent model matrices for any extra validation datasets.
+  xlevels <- .getXlevels(term, df_for_model_matrix)
+
+  # Helper: safely get column names for Matrix and base matrices.
+  get_colnames_safe <- function(mat) {
+    cn <- colnames(mat)
+    if (is.null(cn) && inherits(mat, "Matrix")) {
+      cn <- dimnames(mat)[[2]]
+    }
+    cn
+  }
+
+  # Helper: align a matrix's columns to training columns (add missing columns as zeros, drop extras, reorder).
+  align_matrix_cols <- function(mat, train_cols) {
+    if (inherits(mat, "Matrix")) {
+      cn <- get_colnames_safe(mat)
+      if (is.null(cn)) {
+        stop("Validation matrix has no column names; cannot align to training features.")
+      }
+      common <- intersect(cn, train_cols)
+      mat2 <- mat[, common, drop = FALSE]
+      missing <- setdiff(train_cols, common)
+      if (length(missing) > 0) {
+        zeros <- Matrix::Matrix(0, nrow = nrow(mat2), ncol = length(missing), sparse = TRUE)
+        colnames(zeros) <- missing
+        mat2 <- cbind(mat2, zeros)
+      }
+      mat2 <- mat2[, train_cols, drop = FALSE]
+      mat2
+    } else {
+      cn <- colnames(mat)
+      if (is.null(cn)) {
+        stop("Validation matrix has no column names; cannot align to training features.")
+      }
+      common <- intersect(cn, train_cols)
+      mat2 <- mat[, common, drop = FALSE]
+      missing <- setdiff(train_cols, common)
+      if (length(missing) > 0) {
+        zeros <- matrix(0, nrow = nrow(mat2), ncol = length(missing))
+        colnames(zeros) <- missing
+        mat2 <- cbind(mat2, zeros)
+      }
+      mat2 <- mat2[, train_cols, drop = FALSE]
+      mat2
+    }
+  }
+
+  # Helper: build a LightGBM Dataset from a validation data.frame using the training term/xlevels.
+  build_valid_dataset <- function(df_valid, train_cols, sanitized_cols) {
+    mf_valid <- tryCatch({
+      model.frame(term, data = df_valid, na.action = na.pass, xlev = xlevels)
+    }, error = function(e) {
+      NULL
+    })
+    if (is.null(mf_valid) || nrow(mf_valid) == 0) {
+      return(NULL)
+    }
+
+    y_valid <- model.response(mf_valid)
+    if (length(y_valid) == 0) {
+      return(NULL)
+    }
+    keep_idx <- !is.na(y_valid)
+    mf_valid <- mf_valid[keep_idx, , drop = FALSE]
+    y_valid <- y_valid[keep_idx]
+    if (nrow(mf_valid) == 0) {
+      return(NULL)
+    }
+
+    mat_valid <- if (isTRUE(sparse)) {
```

Comment:
Is 'sparse' parameter covered in the test?

Reply:
Thank you, I added the test case for this.